mace/modules/core_models.py

```diff
@@ -0,0 +1,378 @@
+"""Core model classes, and mixins for various quantities of interest
+
+Notes
+-----
+`MaceCoreModel` is the backbone of MACE models, stores all needed blocks and allows
+for customisable readout shapes.
+
+Mixin classes are defined for calculating quantities:
+- EnergyModelMixin: inherit from if you want to calculate energies
+- ScaleShiftEnergyModelMixin: energies with scaling and shifting
+- DipoleModelMixin: dipole learning
+
+"""
+from typing import Type, List, Optional, Callable, Dict, Tuple
+
+import numpy as np
+import torch
+from e3nn import o3
+from e3nn.util.jit import compile_mode
+
+from mace.modules import (
+    LinearReadoutBlock,
+    NonLinearReadoutBlock,
+    InteractionBlock,
+    LinearNodeEmbeddingBlock,
+    RadialEmbeddingBlock,
+    EquivariantProductBasisBlock,
+    AtomicEnergiesBlock,
+    ScaleShiftBlock,
+)
+from mace.modules.utils import get_edge_vectors_and_lengths, compute_fixed_charge_dipole
+from mace.tools.scatter import scatter_sum
+
+
+@compile_mode("script")
+class MaceCoreModel(torch.nn.Module):
+    """Core model for all MACE models
+
+    Includes the following
+    - graph parameters: r_max, elements, etc.
+    - embeddings (node, edge)
+    - readout blocks (subclasses can change settings of these)
+
+    """
+
+    _LINEAR_READOUT_CLASS = LinearReadoutBlock
+    _NONLINEAR_READOUT_CLASS = NonLinearReadoutBlock
+
+    def __init__(
+        self,
+        r_max: float,
+        num_bessel: int,
+        num_polynomial_cutoff: int,
+        max_ell: int,
+        interaction_cls: Type[InteractionBlock],
+        interaction_cls_first: Type[InteractionBlock],
+        num_interactions: int,
+        num_elements: int,
+        hidden_irreps: o3.Irreps,
+        MLP_irreps: o3.Irreps,
+        avg_num_neighbors: float,
+        atomic_numbers: List[int],
+        correlation: int,
+        gate: Optional[Callable],
+        radial_MLP: Optional[List[int]] = None,
```

Comment:
I have not changed this for the sake of consistency, method parameters are conventionally lower case, so I propose changing these `MLP` parameters to lowercase